### PR TITLE
feature(#687): Implemented xgboost model

### PR DESCRIPTION
**Pull request by egordm**
This pull request introduces several enhancements and new features across the OpenSTEF core and models packages, focusing on improved support for sample weighting and advanced quantile regression capabilities with XGBoost. The main changes include adding support for sample weight columns in validated datasets, updating optional dependencies, and introducing a new utility module for custom loss functions.

**Core feature improvements:**

* Added support for sample weight columns in validated datasets:
  - The `ValidatedDataset` class now accepts a `sample_weight_column` argument and exposes a `sample_weight_series()` method to retrieve sample weights, defaulting to 1 if the column does not exist. The `input_data()` method also excludes the sample weight column from features. [[1]](diffhunk://#diff-d6a52b438eacfa60132a47e31a0b1e96cec8a15dddf32d0dc3bf058eb63a6478R57) [[2]](diffhunk://#diff-d6a52b438eacfa60132a47e31a0b1e96cec8a15dddf32d0dc3bf058eb63a6478R66) [[3]](diffhunk://#diff-d6a52b438eacfa60132a47e31a0b1e96cec8a15dddf32d0dc3bf058eb63a6478R78) [[4]](diffhunk://#diff-d6a52b438eacfa60132a47e31a0b1e96cec8a15dddf32d0dc3bf058eb63a6478R89-R107)

**OpenSTEF models enhancements:**

* Added XGBoost as an optional dependency to enable advanced quantile regression features. (`pyproject.toml`)
* Introduced a new `utils` module for model training support:
  - Added `utils/__init__.py` with documentation and licensing headers.
  - Added `utils/loss_functions.py` implementing custom loss functions for multi-quantile regression with XGBoost, including magnitude-weighted pinball loss, arctan-smoothed pinball loss, and standard pinball loss, all supporting sample weighting.

**Other fixes:**

* Updated the default package name in the `MissingExtraError` exception to `openstef-beam` for consistency.

### CODE REVIEW

File: packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+    )
+    n_jobs: int = Field(
+        default=1, description="Number of parallel threads for tree construction. -1 uses all available cores."
+    )
+    verbosity: Literal[0, 1, 2, 3] = Field(
+        default=1, description="Verbosity level. 0=silent, 1=warning, 2=info, 3=debug"
+    )
+
+
+MODEL_CODE_VERSION = 2
```
Comment (lschilders) [inline packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py]: Why not 1?
```suggestion
MODEL_CODE_VERSION = 1
```

File: packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+    """Serializable state for XGBoost forecaster persistence.
+
+    Contains all information needed to restore a trained XGBoost model,
+    including configuration and the serialized model weights. Used for
+    model saving, loading, and version management in production systems.
+    """
+
+    version: int = Field(default=MODEL_CODE_VERSION)
+    config: XGBoostForecasterConfig
+    model: str
```
Comment (lschilders) [inline packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py]: Should be Pydantic Fields with description. `model: str` is quite vague what it means

File: packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+        return self._config.hyperparams
+
+    @override
+    def to_state(self) -> State:
+        meta: dict[str, Any] = {}
+        meta["_estimator_type"] = self._xgboost_model._get_type()  # noqa: SLF001
+        meta_str = json.dumps(meta)
+        self._xgboost_model.get_booster().set_attr(scikit_learn=meta_str)
+        model_raw = self._xgboost_model.get_booster().save_raw()
+        self._xgboost_model.get_booster().set_attr(scikit_learn=None)
```
Comment (lschilders) [inline packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py:328]: Do you explicitly need the estimator_type thing? When I remove it the tests still succeed. If we require it, we should add a test/assert

File: packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+        )
+
+    @override
+    def predict(self, data: ForecastInputDataset) -> ForecastDataset:
+        if not self.is_fitted:
+            raise NotFittedError(self.__class__.__name__)
+
+        input_data: pd.DataFrame = data.data.drop(columns=[data.target_column])
+        if data.forecast_start is not None:
+            input_data = input_data[input_data.index >= pd.Timestamp(data.forecast_start)]
```
Comment (lschilders) [inline packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py]: We will re-use this everywhere, we can add it in `ForecastInputDataset`

File: packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py (hunk 1 above)
Comment (egordm) [reply packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py]: Woops. A copy paste error. Good find.

File: packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py (hunk 3 above)
Comment (egordm) [reply packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py:328]: Estimator type is only used to checking by xgboost whether the same etimator was used. It is in the original load_model / save_model functions of xgboost. I have kept it. It's not unnecessary but also not that essential.

File: packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py (hunk 4 above)
Comment (egordm) [reply packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py]: Fixed. Movied into input_data(start=)

File: packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py (hunk 2 above)
Comment (egordm) [reply packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py]: Good one. Fixed.
